core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/waiters/WaiterHandlerHelper.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.waiters;
+
+import java.util.List;
+import java.util.Optional;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.retry.RetryPolicyContext;
+import software.amazon.awssdk.core.waiters.PollingStrategy;
+import software.amazon.awssdk.core.waiters.WaiterAcceptor;
+import software.amazon.awssdk.core.waiters.WaiterResponse;
+import software.amazon.awssdk.core.waiters.WaiterState;
+
+/**
+ * The waiter handler helper class
+ */
+@SdkInternalApi
+public final class WaiterHandlerHelper<T> {
```

Comment:
"handler helper". Helps the handler? Seems very generic. Is this just for code sharing between the async and sync code paths?

If so, is there any way we could use a strategy pattern here, where this is a parent class that handles the core logic, but it delegates to the children to do it synchronously or asynchronously?

Reply:
I'm not convinced. That's what the strategy pattern is for. I usually pick composition over inheritance as well, but there are exceptions to the rule. That said, if you've tested it out and it's messy, I believe you. I haven't done the deep-dive necessary to tell whether it's any better than this option.

I notice a lot of duplicate code between the async and sync code paths, though. Can we move some of it to the WaiterHandlerHelper?

Reply:
Are you referring to the switch statement? It's not true duplication. For instance, for exceptional case, sync will throw exception and async will complete the future exceptionally. While it's possible to share more code, the code would be less readable, in my mind unless you have other ideas?

sync:
```java
        if (waiterState.isPresent()) {
            WaiterState state = waiterState.get();
            switch (state) {
                case SUCCESS:
                    return handlerHelper.createWaiterResponse(response, exception, attemptNumber);
                case RETRY:
                    return maybeRetry(pollingFunction);
                case FAILURE:
                    throw SdkClientException.create("A waiter acceptor was matched and transitioned the waiter "
                                                    + "to failure state.");
                default:
                    throw new UnsupportedOperationException();
            }
        }

        if (exception != null) {
            throw SdkClientException.create("An exception was thrown and did not match with any waiter acceptor", exception);
        } else {
            // default to retry if there's no acceptor matched for the response
            return maybeRetry(pollingFunction);
        }
    }
```

async:

```java
            if (waiterState.isPresent()) {
                log.info(() -> "waiter state " + waiterState.get());
                WaiterState state = waiterState.get();
                switch (state) {
                    case SUCCESS:
                        future.complete(handlerHelper.createWaiterResponse(response, exception, attemptNumber));
                        break;
                    case RETRY:
                        maybeRetry(asyncPollingFunction, future);
                        break;
                    case FAILURE:
                        future.completeExceptionally(SdkClientException.create("A waiter acceptor was matched and transitioned "
                                                                               + "the waiter to failure state"));
                        break;
                    default:
                        future.completeExceptionally(new UnsupportedOperationException());
                }
            } else {
                if (exception != null) {
                    future.completeExceptionally(SdkClientException.create("An exception was thrown and did not match any "
                                                                           + "waiter acceptors"));
                } else {
                    // default to retry if there's no acceptor matched for the response
                    maybeRetry(asyncPollingFunction, future);
                }
            }
```

Reply:
I think the sync ("return the result") and async ("return a completed future") behavior can be factored out of the logic itself. It's just a question of how hard it makes the code to read.